userd/settings.go

```diff
@@ -0,0 +1,222 @@
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	if err := settingWhitelisted(setting); err != nil {
+		return "", err
+	}
+	if !allowedSettingsValues.MatchString(check) {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot check setting %q to value %q: value not allowed", setting, check))
+	}
+
+	// FIXME: this works only for desktop files
+	desktopFile := fmt.Sprintf("%s_%s", snap, check)
+
+	cmd := exec.Command("xdg-settings", "check", setting, desktopFile)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot check setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	return strings.TrimSpace(string(output)), nil
+}
+
+// Get implements the 'Get' method of the 'io.snapcraft.Settings'
+// DBus interface.
+//
+// Example usage: dbus-send --session --dest=io.snapcraft.Settings --type=method_call --print-reply /io/snapcraft/Settings io.snapcraft.Settings.Get string:'default-web-browser'
+func (s *Settings) Get(setting string, sender dbus.Sender) (string, *dbus.Error) {
+	if err := settingWhitelisted(setting); err != nil {
+		return "", err
+	}
+
+	cmd := exec.Command("xdg-settings", "get", setting)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot get setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	if !strings.HasPrefix(string(output), snap+"_") {
+		return "NOT_THIS_SNAP.snap.desktop", nil
+	}
+
+	desktopFile := strings.SplitN(string(output), "_", 2)[1]
+	return strings.TrimSpace(desktopFile), nil
+}
+
+// Set implements the 'Set' method of the 'io.snapcraft.Settings'
+// DBus interface.
+//
+// Example usage: dbus-send --session --dest=io.snapcraft.Settings --type=method_call --print-reply /io/snapcraft/Settings io.snapcraft.Settings.Set string:'default-web-browser' string:'chromium-browser.desktop'
+func (s *Settings) Set(setting, new string, sender dbus.Sender) *dbus.Error {
+	if err := settingWhitelisted(setting); err != nil {
+		return err
+	}
+	if !allowedSettingsValues.MatchString(new) {
+		return dbus.MakeFailedError(fmt.Errorf("cannot set setting %q to value %q: value not allowed", setting, new))
+	}
+
+	// see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	new = fmt.Sprintf("%s_%s", snap, new)
+
+	// FIXME: what GUI toolkit to use?
+	// FIXME2: we could support kdialog here as well
+	if !osutil.ExecutableExists("zenity") {
+		return dbus.MakeFailedError(fmt.Errorf("cannot find zenity"))
+	}
```

Comment:
I don't see dependency changes in the packaging; should we add it?

Reply:
I added a recommends, I think we want to support the kde, xfce etc equivalents as well